src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -456,55 +494,43 @@ public string TraceStateString
 
         /// <summary>
         /// If the Activity has the W3C format, this returns the ID for the SPAN part of the Id.  
-        /// Otherwise it returns a zero SpanId. 
-        /// 
-        /// Note that this returns a readonly ref.   This is because SpanId has a cache of the
-        /// Hex string.   However for that cache to be available for subsequent conversions to
-        /// Hex, you have to be operating on the same instance.   Thus if you need the Hex string
-        /// you should store the SpanId in a ref local variable and call the AsHexString property
-        /// which will have the effect of updating Activity's instance so all subsequent uses
-        /// share the same converted string.  
+        /// Otherwise it returns a zero SpanId.
         /// </summary>
-        public ref readonly ActivitySpanId SpanId
+        public ActivitySpanId SpanId
         {
 #if ALLOW_PARTIALLY_TRUSTED_CALLERS
             [System.Security.SecuritySafeCriticalAttribute]
 #endif
             get
             {
-                if (!_spanIdSet)
+                if (!SpanIdSet)
                 {
                     if (_id != null && IdFormat == ActivityIdFormat.W3C)
                     {
-                        _spanId = ActivitySpanId.CreateFromString(_id.AsSpan(36, 16));
-                        _spanIdSet = true;
+                        ActivitySpanId spanId = ActivitySpanId.CreateFromString(_id.AsSpan(36, 16));
+                        _spanId = spanId.ToHexString();
+                        SpanIdSet = true;
+                        return spanId;
```

Comment:
What are the threading requirements for these properties? e.g. if I have an Activity and then from it I create two child Activities, each of which is processed by a separate thread, could I end up in a situation where each of those threads concurrently and legitimately walks up the parent activity chain and accesses this SpanId?

Reply:
Is stored in an AsyncLocal so probably high. 

Use `_spanId` to represent the value and whether its set rather than a separate flag?

Reply:
> Use _spanId to represent the value and whether its set rather than a separate flag?

That would seem to make sense.

Reply:
You've sent me down a rabbit hole... `AddTag` and `AddBaggage` etc

Reply:
> You've sent me down a rabbit hole

If your change isn't making it worse, we could go ahead with it, and then address the threading issue separately.

Reply:
Think I fixed most of them (and may have over compensated)